azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/AzureComputeServiceAdapter.java

```diff
@@ -0,0 +1,345 @@
+   }
+
+   private void getImagesFromPublisher(String publisherName, List<ImageReference> osImagesRef) {
+      OSImageApi osImageApi = api.getOSImageApi(getLocation());
+      Iterable<Offer> offerList = osImageApi.listOffers(publisherName);
+      for (Offer offer : offerList) {
+         Iterable<SKU> skuList = osImageApi.listSKUs(publisherName, offer.name());
+         for (SKU sku : skuList) {
+            osImagesRef.add(ImageReference.create(publisherName, offer.name(), sku.name(), null));
+         }
+      }
+   }
+
+   @Override
+   public Iterable<ImageReference> listImages() {
+      final List<ImageReference> osImages = Lists.newArrayList();
+      getImagesFromPublisher("Microsoft.WindowsAzure.Compute", osImages);
+      getImagesFromPublisher("MicrosoftWindowsServer", osImages);
+      getImagesFromPublisher("Canonical", osImages);
+      return osImages;
+   }
+
+   @Override
+   public ImageReference getImage(final String id) {
+      Iterable<ImageReference> images = listImages();
+      for (ImageReference image : images) {
+         if (id.contains(image.offer()) && id.contains(image.sku())) {
+            return image;
+         }
+      }
+      return null;
+   }
+
+   private String getLocation() {
+      return "eastasia"; // TODO: get location
+   }
+
+   @Override
+   public Iterable<Location> listLocations() {
+      return api.getLocationApi().list();
+   }
+
+   private String getResourceGroupFromId(String id) {
+      String searchStr = "/resourceGroups/";
+      int indexStart = id.lastIndexOf(searchStr) + searchStr.length();
+      searchStr = "/providers/";
+      int indexEnd = id.lastIndexOf(searchStr);
+
+      String resourceGroup = id.substring(indexStart, indexEnd);
+      return resourceGroup;
+   }
+
+   @Override
+   public VMDeployment getNode(final String id) {
+      Deployment deployment = api.getDeploymentApi(getGroupId()).get(id);
+      if (deployment == null)
+         return null;
+      String resourceGroup = getResourceGroupFromId(deployment.id());
+      VMDeployment vmDeployment = new VMDeployment();
+      vmDeployment.deployment = deployment;
+      List<PublicIPAddress> list = getIPAddresses(deployment);
+      vmDeployment.ipAddressList = list;
+      return vmDeployment;
+   }
+
+   @Override
+   public void destroyNode(final String id) {
+      logger.debug("Destroying %s ...", id);
+      String storageAccountName = id.replaceAll("[^A-Za-z0-9 ]", "") + "storage";
+      int index = id.lastIndexOf("-");
+      String group = id.substring(0, index);
+
+      // Delete VM
+      URI uri = api.getVirtualMachineApi(getGroupId()).delete(id);
+      if (uri != null){
+         boolean jobDone = Predicates2.retry(new Predicate<URI>() {
+            @Override public boolean apply(URI uri) {
+               return ParseJobStatus.JobStatus.DONE == api.getJobApi().jobStatus(uri);
+            }
+         }, 60 * 10 * 1000 /* 5 minute timeout */).apply(uri);
```

Comment:
Consider declaring all these predicates in the compute service module so they can be easily reused. Also that will facilitate making the timeouts configurable (they should be). Have a look at the DigitalOcean compute service context module to see an example of how reusable predicates are defined and injected.